docs/04-reference/winglang-spec.md

```diff
@@ -782,30 +782,153 @@ let f = (arg1: num, var arg2: num) => {
-Optionality means the value behind type can be either present or nil.
+Tri-state "nullity" values (`null`, `undefined` in JavaScript) are a primary source of bugs in
+software. Being able to guarantee that a value will never be null makes it possible to write safe
+code without thinking about nullity.
 
-Rules of optionality apply to the entire new container type of `type?` and not
-the value behind it (`type`).  
+So we are going to try and design Wing without allowing users to explicitly assign or use the `nil`
+value. Instead, the language offers syntax to support the various use cases that arise when dealing
+with *optionality*.
 
-Using the `??` operator allows one to safely access the value behind an optional
-variable by always providing a default, in case the variable is nil. This forces
-a value to be present for the l-value (left hand side of the assignment operator
-and guarantees what's returned is type-stripped from its `?` keyword).
+#### 1.7.1 Definition
 
-> ```TS
-> let x? = 44;
-> let y = x ?? 55;
-> ```
+The symbol `?` can mark a variable, function argument, return type or a field as optional (same as
+`Option<T>` in Rust).
 
-<details><summary>Equivalent TypeScript Code</summary>
+```js
+struct Person {
+  name: str;
+  address: str?;
+}
 
-> ```TS
-> const x: number? = 44;
-> const y: number = x ?? 55;
+let x = (foo: str?): num? => {
+  print(foo ?? "foo is not defined");
+  return 12;
+};
+
+class Foo {
+  my_opt: num?;
+}
+```
+
+In the above struct, the `address` field, the `foo` argument and the `my_opt` field are all marked
+as optionals using `?`.
+
+This means, for example, that when a `Person` can be initialized without defining the `address`
+field:
+
+```js
+let my_person = Person { name: "david" };
+```
+
+Let's look at various use cases for how to interact with optionals.
+
+#### 1.7.2 Assignment
+
+A reassignable variable or field can only be assigned to a concrete value. There is no way to
+explicitly assign a "null" value:
+
+```js
+let var x: str?;
+x = "hello"; // OK
+
+x = nil;
+//  ^^^ ERROR: no explicit `nil` in this language.
+```
+
+> You can be a smart-ass and do this:
+> ```js
+> let make_nil = (): str? => { return ?; }
+> x = make_nil();
 > ```
+> Good luck with that.
 
-</details>
+#### 1.7.2 Testing if an optional has a value
+
+Object of type `str?` cannot be used within `==` or `!=` expressions because we don't have a null
+value. Instead, use the the `?` operator to get a `bool` that indicates if the optional is defined
+or not:
+
+```js
+let is_address_defined = my_person.address?; // type is `bool`
+
+// or within a condition
+if my_person.address? {
+  print("address is defined but i do not care what it is");
+}
+
+// can be negated
+if !my_person.address? {
+  print("address is not defined");
+}
+```
+
+#### 1.7.3 Unwrapping within conditions
+
+The `if let` statement can be used to test if an optional is defined an *unwrap* it into a
+non-optional variable defined inside the block:
```

Comment:
I was thinking about the supper common use case of conditionally unwrapping an optional and then using the inner value. Here are some things I things I'd like which I think rust/swift don't do nicely:
1) I want the unwrapped symbol to have the same name as the optional's symbol in my inner scope:
```
let x: num? = 7;
if x? {
  print(x); // The symbol x was implicitly defined to the unwrapped `x` and is shadowing the optional `x` by the `x?` above.
}
```
This of course can't work in all cases, but adding a `let` keyword here still feels too much so:
```
if my_person.address?:address {
  print(address); // address symbol is the unwrapped my_person.address
}
```

2) I want the conditional unwrapping expression to be used in more complex expressions. This can work nicely by combining the `?` operator that returns a boolean with the `if let` functionality:
```
if x? && print_enabled() && my_presoon.address?:address {
  print("${address} ${x}");
}

if x? || print_enabled { // compilation error because there's a code path where we'll enter the `if` without unwrapping
}
```

Reply:
I was trying to avoid this static analysis scenarios as P1

Reply:
Going with `if let` for now to support these use cases without support for complex expressions and without requiring static analysis.